### ISSUE TITLE
Q to quit does not work - just continues

### ISSUE THREAD

**Issue by steve6375** [deomsh/TEXTSTAT.G4B#3]
![image](https://user-images.githubusercontent.com/1469350/203977347-af5595c6-ff52-4f7a-b12f-61bdf86394a3.png)

![image](https://user-images.githubusercontent.com/1469350/203977395-44315ab4-4b54-4238-a566-1d563e727eb6.png)

**Comment by deomsh** [comment]
True, but [Hit Q....] is not part of the script, but provided by 'pager on'. [Hit Q...] never worked (fore me) with pager on, only with cat.
Only escape is to end the script with Ctrl+C 
In case of /P, pager is set on by line 18 (original pager status is available from line 18, and is given back after closing the script in line 87.

**Comment by deomsh** [comment]
My problem with pager is that pager blocks writing keys to the (bios) keyboard buffer, 'head' and 'tail' have always the same value, so interrupt will never exist....

**Comment by deomsh** [comment]
v0.7.2
Interrupt [Hit Q...] is working to abort further printing to screen of lines containing counted 'phrase' and exit script
Highlighting of each 'phrase' inside lines printed to screen

![TEXTSTAT G4B v0 7 2 cmd with COUNT III](https://user-images.githubusercontent.com/67714723/204113165-1c667f76-85cd-40d3-a813-97308fb184b9.jpg)

![TEXTSTAT G4B v0 7 2 new appearance I](https://user-images.githubusercontent.com/67714723/204113095-3ecd72fa-7201-4348-9025-75f1bdbe11ed.jpg)

![TEXTSTAT G4B v0 7 2 after hitting 'Q' II](https://user-images.githubusercontent.com/67714723/204113098-81f247a1-304b-4a47-823d-d4c61370db62.jpg)

**Comment by steve6375** [comment]
great work. :-) 💯 
I suggest that you AND the key code with 0xff so that it also works under grub4dos for EFI (as it does not work at the moment)

e.g.
` set /A keycode=%@retval%&0xff`
and then test for 51 or 71

Also, I would prefer if it did not repeat the same line when using /P - maybe a /P1 switch and if line number was same as previous line number then do not print the line?

**Comment by deomsh** [comment]
Alright, please test version below on grub4dos for UEFI.

BTW can you tell me how grub4dos for UEFI handles keycodes with 'pause --test-key' ? In case of chars seems simple to me, but how about keys like 'Enter' or 'Escape'?

Also switch /P1 added, only two new lines and three additions to setlocal/ endlocal...
Is switch /P1 according to your idea?

![Textstat pagestat g4b count-  -P1](https://user-images.githubusercontent.com/67714723/204137814-423a4bad-e0ba-4f04-b2e6-ce8d88a254da.png)

[TEXTSTAT.ZIP](https://github.com/deomsh/TEXTSTAT.G4B/files/10098450/TEXTSTAT.ZIP)

**Comment by steve6375** [comment]
under UEFI the pause --test-key results are not always the same (especially the top bytes)
Here is ESC and ENTER keys
![image](https://user-images.githubusercontent.com/1469350/204234985-1d28c35a-aa5f-47b4-8208-a74186f67168.png)

New version seems to work under grub4efi for Q for quit now. P1 needs to be added to help text.

seems to be a bug in grub4efi - $[0x70] is not changing background in graphicsmode.

**Comment by deomsh** [comment]
About UEFI key codes: in UEFI specs v2.6 I found a (bit vague) sentence about 16-bits UNICODE characters for keys. The wikipedia article on unicode seems to be a (partly?) corroboration about your suggestion to use `%@retval%&0xff`
Unicode character for 'Enter' is U+000D (bios scancode 0x1C0D) for Carriage return and 'Escape' U+001B for ESC (bios scancode is 0x11B), so last, lowest byte seems 'good'.
Ctrl+A - Ctrl+Z same as lowest byte of bios scancode. Shift+ codes for Latin seem to be same as ASCII on first sight.
Only Alt+ and numlock codes like 0x2100=Alt+F and 0x5032=NumlockOff+Shift+Arrow-Down (0x4838 for Up) ?? BTW I use these in my Text-editor.

About `$[0x70]`-problem in grub4efi: I will try to 'centralize' used highlight-code in a variable, so can be easily changed. Can I assume `$[0x0F]` is working in grub4efi graphicsmode?

About /P1 in 'Help': will be added now, I provided just a test-version in TEXTSTAT.ZIP ('v0.7.3_Vbox' is not a 'ready' version).

**Comment by steve6375** [comment]
YaYa says that grub4efi does not support background colours in graphicsmode. Background colours can only be produced if in text mode under grub4efi in my testing.
This means your color code of $[0x70]  just prints black text on a black background under grub4efi.
Could you change your colours so it uses a black background color and a different foreground color?

Here is Alt+Q, Ctrl+Q and Alt+Ctrl+Q under grub4efi
![image](https://user-images.githubusercontent.com/1469350/204508045-45c50a89-1419-43f1-ad41-7f4a33c2a305.png)

Up = 14000
Down = 25000

**Comment by deomsh** [comment]
Thanks for codes, so Ctrl+ is NOT same as ASCII control-codes

You can try TEXTSTAT_grub4efi.G4B (already try-version 0.8 because of a new switch)

Please FIRST try text-mode - best set BEFORE boot
SECOND graphics-mode (auto-switching background/ foreground color, but maybe you can't go back to text-mode without Reset (if grub4efi has same issue you raised for grub4dos)
[TEXTSTAT_grub4efi.v0.8H_VB.zip](https://github.com/deomsh/TEXTSTAT.G4B/files/10115508/TEXTSTAT_grub4efi.v0.8H_VB.zip)

**Comment by steve6375** [comment]
latest version of grub4efi (BETA) now supports background color - e.g. $[0x70] works now

however, your latest TEXTSTAT_grub4efi.v0.8H_VB.zip does not work in grub4efi  in text mode (graphicsmode 3). OK in graphicsmode -1 800  or 1024.

This seems to be a bug in latest grub4efi but I am not sure where!

**Comment by deomsh** [comment]
Thanks.

In that case a grub4efi version will not be needed.
In readme for next version I will make a remark about grub4efi and graphicsmodes

**Comment by deomsh** [comment]
I made a version which should work in grub4efi text mode too (BOOTx64.EFI only, can't start BOOTIA32.EFI in VBox [Version 6.1.38 r153438 (Qt5.6.2)].
However I had to made dirty provisions for grub4efi text-mode because 'call Fn.4' is not 'good' (in graphics modes okay).

If reading a bigger text file with switch /R in grub4efi, I got a crash after about 40-50 screens. It seems 'clear' is no longer working (all modes!). After next start some error about 'malloc memory'. 

![grub4efi 20221128 graphicsmode 3 TEXTSTAT with switch R gives after x lines error 'out of malloc, memory0 found' V](https://user-images.githubusercontent.com/67714723/206904038-e804dc3a-4ab4-411b-9714-d5c9f91bf437.jpg)

BTW changing used memory areas made no difference (command 'displaymem' used too, I saw no problems with my defaults).
BTW2 switch [/]R overrules switch [/]COUNT

Can not reproduce the problem with writing 40-50 times a few thousand characters to screen and then give command 'clear'.
Maybe the problem has to do with (too many) repetitions of setlocal/ endlocal ?

Test version:

[TEXTSTAT.v0.8ZB_VBox.zip.txt](https://github.com/deomsh/TEXTSTAT.G4B/files/10202194/TEXTSTAT.v0.8ZB_VBox.zip.txt)

**Comment by steve6375** [comment]
I had similar issues with crashing, etc. too. grub4efi is just buggy!
